indico_jacow/plugin.py

```diff
@@ -39,22 +46,61 @@ class JACOWPlugin(IndicoPlugin):
-        self.connect(signals.menu.items, self.add_sidemenu_item, sender='event-management-sidemenu')
-
-    def inject_abstract_export_button(self, event=None):
+        self.template_hook('abstract-list-options', self._inject_abstract_export_button)
+        self.template_hook('contribution-list-options', self._inject_contribution_export_button)
+        self.template_hook('custom-affiliation', self._inject_custom_affiliation)
+        self.connect(signals.core.form_validated, self._form_validated)
+        self.connect(signals.event.abstract_accepted, self._abstract_accepted)
+        self.connect(signals.event.sidemenu, self._extend_event_menu)
+        self.connect(signals.menu.items, self._add_sidemenu_item, sender='event-management-sidemenu')
+        self.connect(signals.plugin.schema_pre_load, self._person_link_schema_pre_load, sender=PersonLinkSchema)
+        self.connect(signals.plugin.schema_post_dump, self._person_link_schema_post_dump, sender=PersonLinkSchema)
+        self.inject_bundle('main.js', WPContributions)
+        self.inject_bundle('main.js', WPDisplayAbstracts)
+        self.inject_bundle('main.js', WPManageAbstracts)
+        self.inject_bundle('main.js', WPManageContributions)
+        self.inject_bundle('main.js', WPMyContributions)
+
+    def _inject_abstract_export_button(self, event=None):
         return render_plugin_template('export_button.html',
                                       csv_url=url_for_plugin('jacow.abstracts_csv_export_custom', event),
                                       xlsx_url=url_for_plugin('jacow.abstracts_xlsx_export_custom', event))
 
-    def inject_contribution_export_button(self, event=None):
+    def _inject_contribution_export_button(self, event=None):
         return render_plugin_template('export_button.html',
                                       csv_url=url_for_plugin('jacow.contributions_csv_export_custom', event),
                                       xlsx_url=url_for_plugin('jacow.contributions_xlsx_export_custom', event))
 
-    def extend_event_menu(self, sender, **kwargs):
+    def _inject_custom_affiliation(self, person=None):
+        return render_plugin_template('custom_affiliation.html', person=person)
+
+    def _form_validated(self, form, **kwargs):
+        if isinstance(form, ContributionForm):
+            person_links = form.person_link_data.data
+            affiliations_cls = ContributionAffiliations
+        elif isinstance(form, AbstractForm):
+            person_links = form.person_links.data
+            affiliations_cls = AbstractAffiliations
+        else:
+            return
+        affiliations_ids = g.pop('jacow_affiliations_ids', {})
+        for person_link in person_links:
+            person_affiliations = affiliations_ids.get(person_link.person.email, [])
+            person_link.jacow_affiliations = []
+            db.session.flush()
+            person_link.jacow_affiliations = [affiliations_cls(affiliation_id=id, display_order=i)
```

Comment:
this is quite ugly but it's a workaround for an error related to the unique constraint i put on `(person_link_id, display_order)`. should I get rid of that constraint altogether?

Reply:
i think it;s fine. there's no other data referencing the jacow affiliations so no problem w/ recreating them while saving. it's all in a transaction anyway